plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/nj2k/conversions/RemoveUnnecessaryParenthesesConversion.kt

```diff
@@ -0,0 +1,213 @@
+// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
+
+package org.jetbrains.kotlin.nj2k.conversions
+
+import org.jetbrains.kotlin.analysis.api.KtAnalysisSession
+import org.jetbrains.kotlin.nj2k.NewJ2kConverterContext
+import org.jetbrains.kotlin.nj2k.RecursiveConversion
+import org.jetbrains.kotlin.nj2k.identifier
+import org.jetbrains.kotlin.nj2k.symbols.JKUniverseClassSymbol
+import org.jetbrains.kotlin.nj2k.symbols.isStaticMember
+import org.jetbrains.kotlin.nj2k.tree.*
+import org.jetbrains.kotlin.nj2k.types.fqName
+import org.jetbrains.kotlin.parsing.KotlinExpressionParsing
+import org.jetbrains.kotlin.utils.addToStdlib.safeAs
+import org.jetbrains.kotlin.utils.filterIsInstanceAnd
+
+/**
+ * Analogous to `RemoveUnnecessaryParenthesesIntention`.
+ * Removes parentheses (conservatively) added by other steps.
+ */
+class RemoveUnnecessaryParenthesesConversion(context: NewJ2kConverterContext) : RecursiveConversion(context) {
+
+    context(KtAnalysisSession)
+    override fun applyToElement(element: JKTreeElement): JKTreeElement {
+        return if (element !is JKParenthesizedExpression || areParenthesesNecessary(element)) {
+            recurse(element)
+        } else {
+            recurse(element::expression.detached().withFormattingFrom(element))
+        }
+    }
+
+    private fun areParenthesesNecessary(expression: JKParenthesizedExpression): Boolean {
+        val innerExpression = expression.expression
+        val parent = expression.parent
+
+        // We can end up with silly "preexisting" parentheses when converting expressions that take significantly different forms in Java
+        // vs. Kotlin. For example, we can have unnecessary parentheses wrapping the left side of an assignment when turning an expression
+        // like `a = b += (c) = ((d *= e))` into `(c) = ((e.let { d *= it; d })) ...`
+        if ((parent is JKKtAssignmentStatement && parent.field == expression) || (parent is JKAssignmentChainLetLink && parent.receiver == expression)) return false
```

Comment:
This is very specific to `AssignmentExpressionUnfoldingConversion`, I would prefer this case to be fixed there (don't add redundant parentheses), if it's not too hard.

Reply:
I don't think it's possible to do this easily for every corner case in `AssignmentExpressionUnfoldingConversion`. For example, when running `AssignmentExpressionUnfoldingConversion` on the java expression `a = b |= (c) = ((d &= e))`, we get to a state in `JKJavaAssignmentExpression.toExpressionChainLink` where the receiver is still `((d &= e))` (and therefore needs a layer of parentheses). Later that binary expression gets transformed into a let expression, but the opportunity to remove the parentheses easily is already gone.

All of that being said, I made a change to `AssignmentExpressionUnfoldingConversion` which covers most cases (e.g. instead of `e.let { d = d and it; d }.also { c = it }.let { b = b or it; b }.also { a = it }` we get `(e.let { d = d and it; d }).also { c = it }.let { b = b or it; b }.also { a = it }`, but that's still better than the original `((e.let { d = d and it; d })).also { c = it }.let { b = b or it; b }.also { a = it }`)